### DISCUSSION TITLE
osu!catch score system issue

### DISCUSSION THREAD

**Discussion by Firuqwq** [ppy/osu#26040] [Q&A]
After the scoring changes made to lazer on June 5th (https://osu.ppy.sh/home/changelog/lazer/2023.605.0 for reference), osu!catch scoring became bad.

I couldn't find exactly how scoring is calculated, so everything I wrote here is based on my own testing and comparing with the score system previous to this change made.

Now to the main issues:

### 1. Low combo plays give too much score due to high accuracy.

Accuracy affects scoring way too much than combo and due to how catch works it's fairly easy to get high accuracy on almost any play you make and even easier on lazer since droplets weigh too little towards accuracy, making 98%+ accuracy easily achievable.

Here is an example, this play was made on April 20th. before the scoring changes.
407/408 droplets, 1 miss and 359/700 combo makes for a total score of 660.000: main factor being not enough high combo.

![imagen](https://github.com/ppy/osu/assets/154583109/9ec32138-d421-4c21-9090-85134cc58fda)

But on this play I made today I got a higher score with a significantly worse play.
406/408 droplets, 11 misses and 108/700 combo makes for a total score of 782.544, which implies combo matters too little towards scoring and accuracy weighs more.
![imagen](https://github.com/ppy/osu/assets/154583109/ff5af9ef-2494-484a-b214-2123cd527982)

### 2. Mods give too much score for bad plays.

Combined with the first issue mods give too much score to bad plays, moded plays are able to bypass on score full combo no mod plays.
This mainly affects spotlights, making the main strategy trying to make mediocre/okay moded plays than full combo with no mods.

On this play I only got 800/4072 combo, but thanks to 99.14% accuracy and Hidden this is worth close to a million points.
Why is this worth almost as a full combo no mod play?
![osu_2023-10-25_14-33-53](https://github.com/ppy/osu/assets/154583109/d4a40a68-f5d6-4bc5-97d7-19c1f50266e6)

Another example, just need to add 2 mods to make a mediocre play worth almost a million.
![osu_2023-12-22_03-16-38](https://github.com/ppy/osu/assets/154583109/602eedc1-80b5-4de2-b941-0c4be491c42e)

### 3. Droplets weigh too little towards accuracy.

On stable; fruits, slider ticks and droplets are worth the same value on accuracy, missing any of them makes you lose the same amount of accuracy,
But on lazer droplets barely lose you accuracy if you miss them, making very easy to achieve high accuracy plays.

On this example I tried to miss as many droplets as possible, the max amount of droplets on this map is 341.

This play was made on stable, geting an expected low accuracy.
![screenshot036](https://github.com/ppy/osu/assets/154583109/4ed50b56-9810-418d-93ed-5b3ff5ff6794)

Then on lazer I made a really similar play, geting too high accuracy for so many missed droplets.
Why missing 76% of the droplets just loses 4.05% accuracy?
![imagen](https://github.com/ppy/osu/assets/154583109/4ce98078-b206-4105-9529-bdb7696c303a)

### How can it be changed?

A great step would be make scoring as similar to stable's score v2 as possible; rolling the scoring back as it was before the change could work but that scoring system has a big flaw of combo altering the final score too much.
As seen on my first example, 1 miss at the middle of the map resulted on -340.000 score, which is too much for only 1 miss.

Other solution could be to find a good balance between combo and accuracy, sadly I can't help with a good example for this.

**Comment by bdach** [comment]
Can you attach replays for these scores that you've been testing with? I want to cross-check against stable and see where the issue is exactly, because I'm not following how "making scoring as similar to stable's score v2 as possible" is not what we've been doing in the update you call broken.

**Reply by Firuqwq** [reply]
Sure, the only replay I don't have is the first one. 
https://drive.google.com/drive/folders/1N07Wk6ZdV9f25o30YB0q8k0JOfz_jdcn?usp=sharing
If needed I can help with more testing.

**Comment by bdach** [comment]
So in order:

> Here is an example, this play was made on April 20th. before the scoring changes.

I cannot and will not address this without the replay present. This score is so old it precedes us bringing stable scoring back. I believe what you see there is old "flat" lazer scoring. It is incomparable with score V2, it was made by us naively thinking we can "streamline" scores across all rulesets which turned out to be a fool's errand as players demanded bringing stable's weird features back.

> But on this play I made today I got a higher score with a significantly worse play.
> 406/408 droplets, 11 misses and 108/700 combo makes for a total score of 782.544, which implies combo matters too little towards scoring and accuracy weighs more.

I checked this play out. After manually splicing it to attach score V2 onto it and playing that replay in stable ([replay here](https://github.com/ppy/osu/files/13778521/Firu.playing.KOTOKO.-.Wing.my.way.ZHSteven.Hell-Jumping.2023-12-22_02-06.%2BSV2.zip)), stable shows 721,743 total score for me with score V2 active. So lazer does not appear to be far off in this respect.

Not making value judgements on this yet, but at least this doesn't appear to be a huge departure from the tried-and-tested stable score V2.

> Mods give too much score for bad plays

I think this likely to be a corollary of scores being too squished together and as such I'm not sure discussing this is productive before getting the baseline nomod score correct.

> Droplets weigh too little towards accuracy.
> On stable; fruits, slider ticks and droplets are worth the same value on accuracy, missing any of them makes you lose the same amount of accuracy,

We can _probably_ emulate this with something like the following (crude and very loosely tested, probably incorrect for bonus objects):

```diff
diff --git a/osu.Game.Rulesets.Catch/Scoring/CatchScoreProcessor.cs b/osu.Game.Rulesets.Catch/Scoring/CatchScoreProcessor.cs
index 4b3d378889..32e2108572 100644
--- a/osu.Game.Rulesets.Catch/Scoring/CatchScoreProcessor.cs
+++ b/osu.Game.Rulesets.Catch/Scoring/CatchScoreProcessor.cs
@@ -32,8 +32,10 @@ protected override double ComputeTotalScore(double comboProgress, double accurac
                    + bonusPortion;
         }
 
+        public override int GetBaseScoreForResult(HitResult result) => result.IsHit() ? 1 : 0;
+
         protected override double GetComboScoreChange(JudgementResult result)
-            => GetBaseScoreForResult(result.Type) * Math.Min(Math.Max(0.5, Math.Log(result.ComboAfterJudgement, combo_base)), Math.Log(combo_cap, combo_base));
+            => base.GetBaseScoreForResult(result.Type) * Math.Min(Math.Max(0.5, Math.Log(result.ComboAfterJudgement, combo_base)), Math.Log(combo_cap, combo_base));
 
         public override ScoreRank RankFromAccuracy(double accuracy)
         {

```

With the above applied, the scores you attached change to:
- the lazer play: 783,187 total score, 52.01% accuracy
- the stable play: 753,458 total score, 52.58% accuracy

The above figures are what is shown _after playing the score back_, not after import. Figures after import may be wrong for several reasons (lazer scores are generally not upgraded to new scoring, stable scores with Score V2 active will be imported without touching total score).

That said, I have not considered broader implications of the above on the upcoming deploy. @peppy @smoogipoo do you think we should be trying to change anything here, or is it water under the bridge now and something we'll potentially fix post-initial-deploy as a second run?

summon @bastoo0 @SecreOsu @kimitakari as potentially interested, feel free to summon more people from your communities as required

**Reply by Kimitakari** [reply]
> This score is so old it precedes us bringing stable scoring back.

Well this is what our request comes from. Stable's Scorev2 in catch is more balanced then the lazer's one because the combo/accuracy split is complete opposite favouring accuracy even more than combo, for no apparent reason.

For mod multipliers, some of us have a theoretical idea on how we can balance the mod multiplier however it still need a discussion before we can confidently lock on these multipliers in action.

**Reply by bdach** [reply]
> Well this is what our request comes from

I think you misunderstand. If the score depicted on that screenshot in the OP was set on lazer, then *that was not score V2*. That was an *entirely different, simplified scoring system*, that has been since ditched since everyone hated it.

> Stable's Scorev2 in catch is more balanced then the lazer's one because the combo/accuracy split is complete opposite favouring accuracy even more than combo, for no apparent reason

That is not true. What _is_ true is that score v2 on stable for catch is a unique snowflake in that it doesn't have a "combo" and an "accuracy" portion. It has a 600k "combo" portion, and a... 400k... "tiny droplets portion"?

Note that this was [discussed at length on the osu!dev discord as far back as *May of this year*](https://discord.com/channels/188630481301012481/542532504478023710/1109031752841838592), and nobody opposed normalising this to just be normal 600k combo/400k accuracy like other rulesets have.

**Reply by SecreOsu** [reply]
the 400k *is* an accuracy portion, it got labeled as a "tiny droplets portion" due to a single miss-communication that somehow still hasn't been fixed yet even though I've tried to clarify that several times now

**Reply by Nirfu** [reply]
There are two big differences between stable and lazer that I'm not sure people were aware of when this was discussed back then.

1. The accuracy calculation which should be fixed with the above suggestion, and therefore the accuracy portion should be as intended with that.
2. As far as I know, tiny droplets contribute to the combo portion in lazer, but not in stable.

If the tiny droplets are removed from the combo portion calculation, the score in lazer should be really similar, if not the same as stable for all FC scores and only slightly different for all other scores, and I think that was the assumption in the discussion back then.

Currently because of the broken accuracy calculation, most scores get almost the entire accuracy portion without much effort.
With the fixed accuracy, it should be closer to stable, but with tiny droplets weighted ~2.5x as much as on stable.

**Reply by bdach** [reply]
> the 400k is an accuracy portion, it got labeled as a "tiny droplets portion" due to a single miss-communication that somehow still hasn't been fixed

it's literally called a "tiny droplets portion" in stable source

**Reply by SecreOsu** [reply]
> it's literally called a "tiny droplets portion" in stable source

im telling you... this was a single miss-communication that NEVER got fixed, when making this change I repeatedly said to the people who were coding that it was accuracy, yet I guess it never got through.

The 400k is ALL accuracy, im the one who helped make the thing lmao I'd hope I know

https://discord.com/channels/188630481301012481/542532504478023710/1109149540352806974

**Reply by bdach** [reply]
```csharp
                double fruitHitsRatio = fruitTinyHitsMaximum == 0 ? 0 : (double)fruitTinyHits / fruitTinyHitsMaximum;

                const int tiny_droplets_portion = 400000;

                return
                    (int)Math.Round
                    ((
                        ((1000000 - tiny_droplets_portion) + tiny_droplets_portion * (1 - fruitTinyScale)) * ComboPortion / ComboPortionMaximum +
                        tiny_droplets_portion * fruitTinyScale * fruitHitsRatio +
                        BonusPortion
                    ) * ModManager.ScoreMultiplier(Score, true));
```

it's not "accuracy". because catch accuracy is

```csharp
        internal override float Accuracy
        {
            get
            {
                if (TotalHits == 0) return 1;

                return (float)TotalSuccessfulHits / TotalHits;
            }
        }

```

it's a separate accuracy-like redefined quantity that does not include fruits or non-small droplets. so i take it you want to have two "accuracies", one for score, one for "accuracy" display, just like the mania people.

sigh. why can nothing ever be easy?

**Reply by SecreOsu** [reply]
as far as I know the calculation can be simplified down to something like:

1000000 - accuracy portion

where accuracy portion = 400000(accuracy)

I *think* this provides the same result, also sorry, forgive me as I'm not well versed in github posts/coding

**Reply by SecreOsu** [reply]
I had a discussion with smoogipoo about this in the past, you can read all the discussion here https://discord.com/channels/188630481301012481/542532504478023710/1109157966814400644

I don't think it ever got sent through because it wasn't a priority at the time to fix 

Making a proper SV2 though will require someone with a broader scope of mathematics and a full team, but using a clone of stable's sv2 should be good enough to be fine for lazer launch I think

**Reply by bdach** [reply]
going back a few:

> As far as I know, tiny droplets contribute to the combo portion in lazer, but not in stable.

i'll need to verify this against stable but *theoretically* this should also be feasible to achieve without further modification to score processor structure.

**Reply by peppy** [reply]
As an update, catch scoring has been rewritten in the latest release, so please give it another try.
